language/diem-vm/parallel_executor/src/scheduler.rs

```diff
@@ -0,0 +1,104 @@
+use crossbeam_queue::SegQueue;
+use crossbeam_utils::CachePadded;
+use std::{
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc, RwLock,
+    },
+};
+
+const FLAG_NOT_EXECUTED: usize = 0;
+const FLAG_EXECUTED: usize = 1;
+
+pub struct Scheduler {
+    // Shared index (version) of the next txn to be executed from the original transaction sequence.
+    execution_marker: AtomicUsize,
+    // Shared number of txns to execute: updated before executing a block or when an error or
+    // reconfiguration leads to early stopping (at that transaction version).
+    stop_at_version: AtomicUsize,
+
+    txn_buffer: SegQueue<usize>, // shared queue of list of dependency-resolved transactions.
+    txn_dependency: Vec<CachePadded<Arc<RwLock<Option<Vec<usize>>>>>>, // version -> txns that depend on it.
```

Comment:
why this is a Option<Vec<usize>>? I see it's only used as Vec<>

Reply:
in order to be able to call replace on option while holding the lock, and then drain the vector without holding the lock. I believe this happens line 83

Reply:
I think you're looking for this replace? https://doc.rust-lang.org/std/mem/fn.replace.html

Reply:
perfect, yes!